### ISSUE TITLE
Dart2js compiler crashes on null values in `while` conditions

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/sdk#45413]
Repro, must be in an opted out library:

```dart
// @dart=2.9
void main() {
  int x = 0;
  do {
    x++;
  } while (x % 2 == 1 ? null : true);
  print(x);
}
```

This stack overflows the compiler:

```
> dart compile js while.dart 
Info: Compiling without sound null safety
Unhandled exception:
Stack Overflow
#0      HInvokeBinary.left (package:compiler/src/ssa/nodes.dart:2500:3)
#1      canBeNegativeZero (package:compiler/src/ssa/invoke_dynamic_specializers.dart:159:36)
#2      ListMixin.any (dart:collection/list.dart:149:15)
#3      canBeNegativeZero (package:compiler/src/ssa/invoke_dynamic_specializers.dart:169:25)
#4      canBeNegativeZero (package:compiler/src/ssa/invoke_dynamic_specializers.dart:159:12)
#5      ListMixin.any (dart:collection/list.dart:149:15)
#6      canBeNegativeZero (package:compiler/src/ssa/invoke_dynamic_specializers.dart:169:25)
#7      canBeNegativeZero (package:compiler/src/ssa/invoke_dynamic_specializers.dart:159:12)
#8      ListMixin.any (dart:collection/list.dart:149:15)
...
```

cc @rakudrama

**Comment by jakemac53** [comment]
@vsmenon pointed out this also repros if you substitute `false` for `null` in the condition, so this isn't specifically related to `null` and that probably does also repro in an opted in library.

**Comment by rakudrama** [comment]
Fix ready for review: https://dart-review.googlesource.com/c/sdk/+/192606

**Comment by rakudrama** [comment]
Fixed via 93313bf12d88a43ac6afcc51f99c729ed7cfa36f
